### ISSUE TITLE
Enhancement for scheduler perf

### ISSUE THREAD

**Issue by draveness** [kubernetes/kubernetes#81719]
/priority important-longterm
/sig scheduling
/assign

Currently, the scheduler perf works pretty well for most of the time, but some issues could affect the benchmark numbers. We could improve the accuracy and usability through the following tasks:

- [x] Get a precise benchmarking test result for each test
    - [x] Remove the cost of the `defer` function in the benchmark scheduling https://github.com/kubernetes/kubernetes/pull/82330
- [x] Use watch apiserver instead of listing all the pods and remove the `time.Sleep` in benchmark scheduling tests https://github.com/kubernetes/kubernetes/pull/82338
- [ ] Evaluate the semantics of b.N in Golang, and use it properly in the benchmark tests. Right now we manually set b.N to the pods' number of a test case https://github.com/kubernetes/kubernetes/issues/84306
- [ ] Consider making the workload customizable and easy to use, e.g. pass workload YAMLs instead of hard-coded specs https://github.com/kubernetes/kubernetes/issues/84307

@bsalamat do you have any thoughts on improving the current scheduler perf?

Thank @Huang-Wei for coaching and pointing out issues in the current scheduling performance tests.

**Comment by Huang-Wei** [comment]
> Remove the cost of the defer function in the benchmark scheduling

Per our offline discussion, we'd like to get a precise benchmarking result for each test, so it includes several things like concise usage of `defer()` as well as other ones like `b.ResetTimer()`, etc.

Could you please rephrase it?

> Use the default b.N instead of setting it manually in the benchmark tests

How about rewording to: Evaluate the semantics of `b.N` in Golang, and use it properly in the benchmark tests. Right now we manually set `b.N` to the pods number of a test case".

**Comment by ahg-g** [comment]
Another line of work related to this issue is improving benchmark coverage. We could create another umbrella issue to track the list of features that are currently not benchmarked and make sure we cover them, such as pod affinity (the part implemented as a priority), even pod spreading etc. This will greatly help us as we migrate to the framework.

**Comment by draveness** [comment]
> Another line of work related to this issue is improving benchmark coverage. We could create another umbrella issue to track the list of features that are currently not benchmarked and make sure we cover them, such as pod affinity (the part implemented as a priority), even pod spreading etc. This will greatly help us as we migrate to the framework.

SG, I'll update the description.

**Comment by ahg-g** [comment]
We have another two items that should be on the list:
1) Improve coverage, for example add interpod affinity benchmarks for PreferredDuringSchedulingIgnoredDuringExecution: https://github.com/kubernetes/kubernetes/pull/82841#discussion_r326191534
2) Fix the dashboard: #84186

**Comment by Huang-Wei** [comment]
2 more added asking for help:

- Evaluate b.N in scheduler's benchmark tests #84306
- Refactor scheduler benchmark tests to make it more extensible #84307

**Comment by liu-cong** [comment]
Another idea: Unify dashboard/metrics for benchmark and clusterloader tests, see [comment](https://github.com/kubernetes/kubernetes/issues/84306#issuecomment-558198172)

**Comment by liu-cong** [comment]
I have recently discussed/experimented with @ingvagabund about scheduler benchmark test. Here is a summary of the current status of this issue as well as proposals on how to improve it.

## b.N in scheduler benchmark tests, ref  #84306
**Proposal**: Keep `b.N` as is.
I don't think it's wrong to set `b.N` to `minPods`. We do this as a "shortcut" to skip initial attempts of `testing.b` with small `b.N`. We could avoid doing this, but this will results in additional attempts which involves setting up/tearing down the cluster that only makes the tests run longer.

## Make benchmark tests configurable  #84307
We should be able to configure the cluster setup(nodes, pods, etc.) as well as the pod spec to schedule. This allows us to easily add new test cases to cover more features.
See #85662 for a PoC.

## Test metrics collector
We should allow collecting both default scheduler metrics (Prometheus) and custom metrics easily. At a high level, these collectors should produce a json formatted result which will then be consumed by perf-dash to present in the dashboard. 

### Default Prometheus metrics
We should provide a metrics collector that collects and summarizes default scheduler metrics defined in [metrics.go](https://github.com/kubernetes/kubernetes/blob/8995c1e030643f26cad45fa33004d86726a2911d/pkg/scheduler/metrics/metrics.go#L23). 
To begin with, we can manually pick a list of interesting metrics. Longer term, we could allow configuring the metric list in test definition. @ingvagabund already had a PoC #85455.
As a reference, [here](https://github.com/kubernetes/perf-tests/blob/44971cd21c3a33916402bc13b4ec049d097e9263/clusterloader2/pkg/measurement/common/scheduler_latency.go#L17) is how clusterloader collects scheduler latency metrics.

### Custom test metrics
The most important metric is scheduling throughput. [Here](https://github.com/kubernetes/perf-tests/blob/2853ca27d2c76e453188019023705d44c3fa28f7/clusterloader2/pkg/measurement/common/scheduling_throughput.go#L30) is how clusterloader calculates the throughput. We can start with just calculating the average throughput without the percentiles.
We may want other custom metrics that are either common to all tests or specific to certain tests. We should allow these custom collectors to be added easily and ideally configurable.

## Test coverage
Once we make the test configurable, we can easily add more test cases to cover more scheduler features.

# What's next
To explain the ideas better, I created a draft PR #85662 that shows a PoC of how to configure the test as well as metrics collection. 
@ingvagabund already has a PoC #85455 of how prometheus metrics can be collected during the benchmark test. Would you like to continue implementing the metrics collector?
@notpad started the effort to configure benchmark test via YAML #84418. Would you like to extend that to accommodate for the more complicated configurations?

**Comment by draveness** [comment]
> I don't think it's wrong to set b.N to minPods. We do this as a "shortcut" to skip initial attempts of testing.b with small b.N. We could avoid doing this, but this will results in additional attempts which involves setting up/tearing down the cluster that only makes the tests run longer.

The golang testing tools allow us to set `b.N` in the benchmark test cases, but I can't say the right or wrong of this approach. It does skip the initial attempts with small `b.N`. However, it could also cost more time than expected after the time of each iteration returns to a stable value. 

> We should be able to configure the cluster setup(nodes, pods, etc.) as well as the pod spec to schedule. This allows us to easily add new test cases to cover more features.

👍

> We should allow collecting both default scheduler metrics (Prometheus) and custom metrics easily. At a high level, these collectors should produce a json formatted result which will then be consumed by perf-dash to present in the dashboard.

Adding metrics to the scheduler perf test is fine; it just not the first intent to create this issue. 

> The current perf-dash has many dependencies such as Prow, GCS, etc. It will be great if we can set up a local dashboard without these dependencies. @ingvagabund tried to modify perf-dash to point to local files in gist to prove it works.

Running a lot of perf tests locally and display the result sounds good, but it could take quite a lot of time. If it would take more than 10 minutes to get the numbers on the PC, the experience could be terrible. Is it possible to run the perf tests in the CI checks?

And thanks @liu-cong for your work!

**Comment by liu-cong** [comment]
> Running a lot of perf tests locally and display the result sounds good, but it could take quite a lot of time. If it would take more than 10 minutes to get the numbers on the PC, the experience could be terrible. 

The intent of the local dashboard is to give developers an alternate tool at analyzing scheduler performance and not to run every time a new code change is done. For instance, when you develop a new plugin and want to do some comprehensive evaluation.

But anyway, this is just a nice to have.

> Is it possible to run the perf tests in the CI checks?

We can consider this. However, it's probably tricky to come up with a criteria to fail the check as the results can be fluctuating. That said, we can do something like taking average of last 10 runs, and make sure the new code doesn't exceed, say, 20% of the average. Let me create an issue for this.

**Comment by draveness** [comment]
> We can consider this. However, it's probably tricky to come up with a criteria to fail the check as the results can be fluctuating. That said, we can do something like taking average of last 10 runs, and make sure the new code doesn't exceed, say, 20% of the average. Let me create an issue for this.

We could make the test optional and only triggered when we reply something like `/test scheduling- performance`. And the test won't be the blocker of the PR process.

**Comment by Huang-Wei** [comment]
@liu-cong Thanks for working on this. The direction looks generally good to me. One thing confused me a bit is: are you trying to use prometheus metric to measure the benchmark? (In terms of benchmark, it's meant to be the `*testing.B` offered by Golang testing framework). If it's the case, incorporating prometheus metric into benchmark tests seem to result in 2 sets of metrics (Golang and Prometheus). How are you gonna deal with them?

**Comment by liu-cong** [comment]
@Huang-Wei 
First of all, we do not necessarily need to use golang benchmark any more. Given that we are updating benchmark tests to run only medium and large cluster #85808,  we can get the same (and more) benchmark metrics as golang benchmark.
Moreover, we need more intuitive metrics than `ns/op`.  We should at least have throughput as well as latency metrics (which helps us identify which step is taking longer). This also aligns with the clusterloader tests.

On a side note: the metrics are just some data saved in logs which is later on parsed by the perf dashboard. We just need to make sure the parser can understand the metrics we are interested in.

**Comment by Huang-Wei** [comment]
> we do not necessarily need to use golang benchmark any more

Then my suggestion is to create a new set of "performance" tests which has nothing to do with `testing.B` - as you said, "performance" tests has a larger scope than the golang "benchmark" test. If we modify current benchmark tests in place, it would cause confusion and doesn't make sense since the output generated by Golang benchmark tests is pointless.

---

Actually my original thought is to leverage the "cluster setup" assets in scalability tests (i.e. clusterloader v2) to make current benchmark tests more flexible (#84307). And as they're Golang benchmark tests, so adhering to Golang test principles is critical, which is why I brought up #84306 .

Now it seems (@ingvagabund correct me if I'm wrong ) #84306 is hard to achieve, then we may have to, as you suggested, alter the direction to create in-house "performance" tests which leverages prometheus metrics and clusterloader v2.

**Comment by ingvagabund** [comment]
> Now it seems (@ingvagabund correct me if I'm wrong ) #84306 is hard to achieve

You are correct. `b.N` as is used now is always set to `minPods` so the golang benchmark does not really adjust `b.N` based on the last `ns/op` to find "optimal" number of iterations. Even when I remove the line overwriting `b.N`, measured `ns/op` values ran 10 times over the same number of nodes/pods and the same commit I get different results which are far from asymptotic. I assumed the longer we run each iteration, all `ns/op` number would be closer to each other. They are not.  I have also increased the benchtime to e.g. 20s or higher to see if there's any improvements. I have not been able to observe any.

Also, even with metrics I get quite a lot of fluctuations opposed to expected wave line parallel to x axis.

**Comment by liu-cong** [comment]
@Huang-Wei , @ingvagabund 
Yeah I am now inclined to defining our "in-house" performance tests and getting rid of golang benchmark. The benchmark gives little value but adds confusion.

> Actually my original thought is to leverage the "cluster setup" assets in scalability tests (i.e. clusterloader v2) to make current benchmark tests more flexible (#84307). And as they're Golang benchmark tests, so adhering to Golang test principles is critical, which is why I brought up #84306 .

I was thinking to use the clusterloader2 config as well. However, after taking a quick look, that seems overkill for us. There are lots of concepts that scheduler tests don't necessarily need. I was wondering if you have something concrete in mind that I might have overlooked. Otherwise I prefer to define our own configuration, which turned out to be very simple and flexible. See #85662.

**Comment by Huang-Wei** [comment]
> Yeah I am now inclined to defining our "in-house" performance tests and getting rid of golang benchmark. The benchmark gives little value but adds confusion.

It makes sense. Then we can deprecate the old "BenchmarkXYZ" tests when new ones are ready.
BTW: in new tests, we are not gonna use `testing.B` or `testing.T`, correct? And I suggest we name the new perf tests starting with "PerfXYZ" to distinguish with Golang UT (`testing.T`) and Benchmark test (`testing.B`).

> Otherwise I prefer to define our own configuration

Sure, please go ahead.

**Comment by Huang-Wei** [comment]
> Even when I remove the line overwriting `b.N`, measured `ns/op` values ran 10 times over the same number of nodes/pods and the same commit I get different results which are far from asymptotic. I assumed the longer we run each iteration, all `ns/op` number would be closer to each other. They are not.

Thanks @ingvagabund for confirming. How different (percentage) is the result? (I suppose the tests are based on "automatic" adjustment on `b.N`, right? What's the number of internal running cycles each time?)

**Comment by liu-cong** [comment]
>BTW: in new tests, we are not gonna use testing.B or testing.T, correct? 
Definitely not `testing.B` any more. But I guess we can use `testing.T` so we can leverage the existing test script to invoke the tests. Any objection with this?

> How different (percentage) is the result? (I suppose the tests are based on "automatic" adjustment on b.N, right? What's the number of internal running cycles each time?)

I checked this before. I noticed `b.N` was adjusted a few times for smaller tests (100 Pods to schedule for instance) but definitely not for the 1000 pods case. Since we removed the small test, `b.N` has no effect any more.
I also created a timer and calculate `ns/op` by dividing the time over number of pods, and it's the same with the result provided by golang benchmark.

**Comment by Huang-Wei** [comment]
> Any objection with this?

Let's see how it goes. Generally, it should be fine.

**Comment by ahg-g** [comment]
Thanks @liu-cong for the detailed plan, the direction looks good to me. Regarding metrics collection, one additional metric we could report is the distribution of node utilization of different resources (cpu, memory and number of pods), this will help understand how the different benchmarks assign pods to nodes.

I agree that we should keep use testing.T so that we can run it as part of integration tests.

**Comment by ingvagabund** [comment]
> > BTW: in new tests, we are not gonna use testing.B or testing.T, correct?

> Definitely not testing.B any more. But I guess we can use testing.T so we can leverage the existing test script to invoke the tests. Any objection with this?

> I agree that we should keep use testing.T so that we can run it as part of integration tests.

Anything under `test/integration` utilizing `testing.T` is ran as part of `make test-integration` execution. Advantage of any test utilizing `testing.B` is it's not run as part of the execution. Given benchmark tests are utilized to measure time, resource consumption, throughput, etc. running them into completion does not give us any value compared to `testing.T` tests. So if we utilize `testing.T`, we need to make sure they are not run as part of `make test-integration`. If we keep `testing.B` and ignore the benchmark output, we can still generate data item files with various metrics (as is done in clusterloader tests these days). Plus, we may update the scheduler performance Jenkins job to run only a subset of benchmark tests under `scheduler_perf` directory. 
I have already updated `hack/jenkins/dockerized.sh` script to allow to override go test's `-bench` option: https://github.com/kubernetes/kubernetes/pull/85915. The idea is to have the current Jenkins job run the current benchmark tests (prefixed with `BenchmarkScheduling`) and provide new tests (https://github.com/kubernetes/kubernetes/pull/85861) in the same directory. Additionally, we will create a new Jenkins Job that will mirror what we have for the scheduler perf but will run benchmark tests with different prefix (e.g. `BenchmarkPerfScheduling`). This way, we can keep the original suite as it is and deprecate it eventually once the new suite gives at least as good results as the current one.

**Comment by ahg-g** [comment]
/close

The evaluation framework is in place now.
